File: Arrays/493-Reverse-Pairs/JavaScript/493-Reverse-Pairs-Brute Force.js
/**
 * Problem: 75. Sort Colors
 * Approach: Better - Counting
 *
 * Count the number of 0s, 1s, and 2s, then overwrite the array
 * with the required number of each value.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * @param {number[]} nums
 * @return {void}
 */

var sortColors = function(nums) {
    let count0 = 0;
    let count1 = 0;
    let count2 = 0;

    // Count 0, 1 and 2
    for (const num of nums) {
        if (num === 0) {
            count0++;
        } else if (num === 1) {
            count1++;
        } else {
            count2++;
        }
    }

    // Fill 0s
    for (let i = 0; i < count0; i++) {
        nums[i] = 0;
    }

    // Fill 1s
    for (let i = count0; i < count0 + count1; i++) {
        nums[i] = 1;
    }

    // Fill 2s
    for (let i = count0 + count1; i < nums.length; i++) {
        nums[i] = 2;
    }
};

File: Arrays/493-Reverse-Pairs/Python/493-Reverse-Pairs-Brute Force.py
"""
Problem: 493. Reverse Pairs
Approach: Brute Force

Check every pair (i, j) where i < j.
A reverse pair exists when:
    nums[i] > 2 * nums[j]

Time Complexity: O(n^2)
Space Complexity: O(1)
"""

class Solution:
    def reversePairs(self, nums: list[int]) -> int:
        n = len(nums)
        count = 0

        for i in range(n):
            for j in range(i + 1, n):
                if nums[i] > 2 * nums[j]:
                    count += 1

        return count